src/accountsdb/cache.zig

```diff
@@ -0,0 +1,365 @@
+        account: Account,
+        /// represents whether this account has been mutated, and shouldn't progress to newer slots
+        // TODO: when we start mutating accounts, make sure to set this field
+        is_dirty: std.atomic.Value(bool) = .{ .raw = false },
+        ref_count: std.atomic.Value(usize),
+
+        pub fn init(allocator: std.mem.Allocator, account: Account) !CachedAccount {
+            return .{
+                .account = try account.clone(allocator),
+                .ref_count = std.atomic.Value(usize).init(1),
+            };
+        }
+
+        pub fn deinit(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            self.account.deinit(allocator);
+            self.* = undefined;
+        }
+
+        pub fn releaseOrDestroy(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            const current_count = self.ref_count.load(.acquire);
+            if (current_count == 1) {
+                self.deinit(allocator);
+                allocator.destroy(self);
+            } else {
+                _ = self.ref_count.fetchSub(1, .acq_rel);
+            }
+        }
+
+        // satisfies the type LruCacheCustom expects
+        pub fn releaseOrDestroyDoublePtr(self: **CachedAccount, allocator: std.mem.Allocator) void {
+            releaseOrDestroy(self.*, allocator);
+        }
+
+        pub fn copyRef(self: *CachedAccount) *CachedAccount {
+            _ = self.ref_count.fetchAdd(1, .acq_rel);
+            return self;
+        }
+
+        pub fn getMeta(self: *CachedAccount) CachedAccountMeta {
+            return .{
+                .is_dirty = self.is_dirty.load(.acquire),
+                .ref_count = self.ref_count.load(.acquire),
+            };
+        }
+    };
+
+    /// A copy of a CachedAccout's fields, without the account
+    pub const CachedAccountMeta = struct {
+        is_dirty: bool,
+        ref_count: usize,
+    };
+
+    pub const LRU = LruCacheCustom(
+        .locking,
+        Pubkey,
+        *CachedAccount,
+        std.mem.Allocator,
+        CachedAccount.releaseOrDestroyDoublePtr,
+    );
+    pub const SlotLRUs = std.AutoHashMapUnmanaged(Slot, LRU);
+
+    const Self = @This();
+
+    slot_lrus: SlotLRUs,
+    allocator: std.mem.Allocator,
+    max_items: usize,
+    max_slots: usize,
+    highest_slot: ?Slot,
+
+    pub fn init(
+        allocator: std.mem.Allocator,
+        max_items: usize,
+        max_slots: usize,
+    ) !AccountsCache {
+        return .{
+            .slot_lrus = .{},
+            .allocator = allocator,
+            .max_items = max_items,
+            .max_slots = max_slots,
+            .highest_slot = null,
+        };
+    }
+
+    /// user is expected to release the returned CachedAccount
+    pub fn get(self: *const Self, slot: Slot, pubkey: Pubkey) ?*CachedAccount {
+        const slot_lru = self.slot_lrus.getPtr(slot) orelse return null;
+        return (slot_lru.get(pubkey) orelse return null).copyRef();
```

Comment:
Maybe split this into a 
```zig
const account = slot_lru.get(pubkey) orelse return null;
return account.copyRef();
```
I feel like that would look cleaner. Same for the one below.